### DISCUSSION TITLE
Wall mode with filter by thread name

### DISCUSSION THREAD

**Discussion by toktarev** [async-profiler/async-profiler#539] [Q&A] [answered]
Hello.

We have applications with hundreds of threads.
However, most of them sleep almost all the time and it is interesting to analyze only 6 active from them.

Is it possible to profile only events from threads whose name can be matched by regexp or substring for example?

**Comment by krzysztofslusarski** [comment]
I don't know if there is such a feature in async-profiler. You can fetch profile output and post-process it:

1. JFR
- Use JFR output
- Convert JFR to collapsed stack (for example with my tool https://github.com/krzysztofslusarski/collapse-jfr)
- ```grep``` the collapsed stack with thread name
2. Collapsed stack with ```-t``` option (or adding ```threads``` option from Java API) and do ```grep```
3. Open JFR format in some tool (like IntelliJ Idea) where you can filter by thread

**Reply by toktarev** [reply]
Will the async profiler loose some accuracy if final jfr contains data for all threads (not only for active) ?

**Comment by apangin** [comment, marked as answer]
There are multiple ways to filter threads in wall clock mode:
1. Use `filter` option. However, it does not accept thread names, but rather thread IDs, e.g `thread=123,140-150`. IDs can be decimal or hex, like `0xaf5`. `jstack` may help to find ID by a thread name:
```
"http-nio-8080-exec-1" #85 daemon prio=5 os_prio=0 tid=0x00007f3d43388800 nid=0x3bf7 waiting on condition [0x00007f3ce5ffa000]
                                                                              ^^^^^^
```
When `filter` is used, other threads won't be sampled at all.

2. Since you already use Java API to start the profiler, it may be handy to filter with Java API, too. Start the profiler with an empty `filter` option, then call [`AsyncProfiler.addThread`](https://github.com/jvm-profiling-tools/async-profiler/blob/5f37bf3ad6e77a692c80985f715548d8dd1c13b4/src/api/one/profiler/AsyncProfiler.java#L183) to include the given thread(s) in the profiling.
3. Use `include`/`exclude` options together with `threads` to filter threads during dumping. E.g. `include=http*,exclude=*BlockPoller, threads` will dump samples from all threads starting with `http`, but not ending with `BlockPoller`.
Note: these options filter samples only during dumping; samples from other threads will be still collected.
4. Use `jfr` output, then open results with JMC or IDEA, where you can filter threads in UI.

**Reply by toktarev** [reply]
Thanks a lot. These are really very good advices.
 
But I will let myself some discussion about usability, of course, 
I don't ask someone to implement something, just discussion.

1. I don't have some ids only thread name  like:
[Executor task launch worker for task 0.0 in stage 36.0 (TID 8954) tid=5014]

2. I don't create those threads, Spark creates them.

3. Of course it works and looks good, but I have more than 100 other dummy threads,
do we really need to collect them and consume a lot of memory for nothing?

4. Sure jfr is a good option but, the same question 
do we really need to collect them and consume a lot of memory for nothing?

So I think that filtering by thread/cpu would be a create improvements for the profiler.
I know that there is some PR regarding CPU filtering (which hungs), and suspect that
filtering by thread name on the collection phase is not so difficult task.

**Reply by toktarev** [reply]
+ I think that avoiding 90% of traffic during collections would let to collect really important information with higher precision.

**Reply by krzysztofslusarski** [reply]
> * I think that avoiding 90% of traffic during collections would let to collect really important information with higher precision.

I believe it is not the case. Correct me @apangin if I'm wrong, but filtering threads will not increase precision. It is constant. It may reduce overhead if you can filter it before calling ```AsyncGetCallTrace```.

**Reply by apangin** [reply]
Exactly. Thanks @krzysztofslusarski

**Reply by apangin** [reply]
BTW, there is an open issue on this topic: #237

**Reply by toktarev** [reply]
Let's imagine we have sampling timeout = 1 ms.

Calling AsyncGetCallTrace for 1000 threads with such little interval could lead to the
cases when we miss some events (just because we don't have time to process
1000 thread having such little timeout).
On the other hand, by having 6 threads we could avoid such samples misses.

Maybe I do not correctly realize what exactly happens internally, please
correct me if I am wrong.

**Reply by toktarev** [reply]
@apangin 
Why did you close https://github.com/jvm-profiling-tools/async-profiler/pull/246 ?

I realize that PR contains a lot of problems but probably we could fix them and merge ?

**Reply by apangin** [reply]
It was submitted before `filter`, `addThread/removeThread` and `include/exclude`  features, and wasn't implemented properly.
By now, it diverged too far from the mainline. You may submit a new PR if you wish.

**Reply by toktarev** [reply]
From what I understand `addThread/removeThread` already implements some filtering on the collections phase?

So we need just some configuration-level improvements?

**Reply by apangin** [reply]
Yes, `filter` option performs filtering during collection. `addThread/removeThread` methods allow changing this filter in runtime.
